models/auth/secret.go

```diff
@@ -0,0 +1,60 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package auth
+
+import (
+	"fmt"
+	"regexp"
+
+	"code.gitea.io/gitea/models/db"
+	"code.gitea.io/gitea/modules/timeutil"
+)
+
+type ErrSecretNameInvalid struct {
+	Name string
+}
+
+func (err ErrSecretNameInvalid) Error() string {
+	return fmt.Sprintf("secret name %s is invalid", err.Name)
+}
+
+type ErrSecretDataInvalid struct {
+	Data string
+}
+
+func (err ErrSecretDataInvalid) Error() string {
+	return fmt.Sprintf("secret data %s is invalid", err.Data)
+}
+
+var nameRE = regexp.MustCompile("[^a-zA-Z0-9-_.]+")
+
+// Secret represents a secret
+type Secret struct {
+	ID          int64
+	UserID      int64              `xorm:"index NOTNULL"`
+	RepoID      int64              `xorm:"index NOTNULL"`
+	Name        string             `xorm:"NOTNULL"`
+	Data        string             `xorm:"TEXT"`
+	PullRequest bool               `xorm:"NOTNULL"`
```

Comment:
This column indicates whether a pull request could read the secret. I think we could have two types `CommitType` and `PullRequestType` according your thought.

Reply:
We may use secrets outside of the git scope. A `PullRequest` column looks wrong then. The same could be said about the `RepoID`.

Reply:
I agree with the `PullRequest`.
It seems like we need a more advanced access control system.
However, I disagree with the `RepoID`, as if it is not present, it is simply an org secret instead of a repo secret, and a repo secret is still an often needed thing.
Our highest-level concepts are still users/orgs and repos (and issues to some extent, but they are irrelevant here). All of them are included in the current implementation…

Reply:
I think we don't need to add `Type` field for secrets, the `PullRequest` is't a type, it indicates whether it can be used in CICD tasks triggered by pull requests, so I prefer something like:
```go
type Secret struct {
	// ...
	ReadByPullRequest bool
	ReadByExternalAPI bool // for example
	ReadByXXX bool // then add as needed
}
```

Reply:
Hey, let's copy someone's homework, maybe the field is unnecessary.

<img width="801" alt="image" src="https://user-images.githubusercontent.com/9418365/206675179-5a9317ba-634c-40c8-9987-8e81620b92f1.png">

See: [Using encrypted secrets in a workflow](https://docs.github.com/en/actions/security-guides/encrypted-secrets#using-encrypted-secrets-in-a-workflow)